lighter/callbacks/writer/base.py

```diff
@@ -1,146 +1,177 @@
-            # Iterate through each prediction for the `idx`-th input sample.
-            for identifier, tensor in parsed_output.items():
-                # Save the prediction in the specified format.
-                self.write(idx, identifier, tensor.detach().cpu(), self.parsed_write_format[identifier])
-
-    def _parse_write_format(self, write_format, parsed_outputs: Dict[str, Any]):
-        # If `write_format` is a string (single value), all outputs will be saved in that specified format.
-        if isinstance(write_format, str):
-            parsed_write_format = {key: write_format for key in parsed_outputs}
-        # Otherwise, `write_format` needs to match the structure of the outputs in order to assign each tensor its type.
+        It decollates the predicted outputs and, if provided, the associated IDs.
+        If the IDs are not provided, it generates global unique IDs based on the prediction count.
+        Finally, it writes the predictions according to the specified format.
+        """
+        # Fetch and decollate preds.
+        preds = decollate_batch(outputs["pred"], detach=True, pad=False)
+        # Fetch and decollate IDs if provided.
+        if outputs["id"] is not None:
+            ids = decollate_batch(outputs["id"], detach=True, pad=False)
+        # Generate IDs if not provided. An ID will be the global index of the prediction.
         else:
-            parsed_write_format = parse_data(write_format)
-            if not set(parsed_write_format) == set(parsed_outputs):
-                logger.error("`write_format` structure does not match the prediction's structure.")
-                sys.exit()
-        return parsed_write_format
+            ids = []
+            for _ in range(len(preds)):
+                # Append the current prediction count to the IDs list.
+                ids.append(self._pred_count)
+                # Increment the prediction count by the total number of DDP processes.
+                # This ensures each process will generate unique IDs in the next batch.
+                self._pred_count += trainer.world_size
+
+        # Iterate over the predictions and save them.
+        for id, pred in zip(ids, preds):
+            # Convert predictions into a structured format suitable for writing.
+            pred = flatten_structure(pred)
+
+            # If a single format is provided, assign it to all pred keys. Otherwise, the format must match the pred structure.
+            format = {key: self.format for key in pred} if isinstance(self.format, str) else flatten_structure(self.format)
+            # Ensure that the format structure matches the prediction structure.
+            if not set(format) == set(pred):
+                raise ValueError("`format` structure does not match the prediction's structure.")
+
+            # If pred is multi-output, there will be a `multi_pred_id` for each output.
```

Comment:
I find this `multi_pred_id` very confusing naming. For the `id` these are some unique identifiers so it makes sense. 

But the `multi_pred_id` is essentially `pred` keys right? So they are repeating for every prediction and the `id` makes me think differently. 

I feel like it would be simpler if to `self.write` we passed the pred, id and format, so
`self.write(pred, id, format)` and then the write functions would handle how to deal with the keys. This should provide more flexibility.

For example, if I had a model with two different heads to predict scalar values and wanted to save the output as 
a single `dict`, I can't do it here because the base class will separate my outputs by default.

Reply:
Hmm, I see the point. I do think that that would require a lot of logic in any `self.write()` that wouldn't necessarily be there.

How about, instead of doing that, to have an argument at `__init__` like `atomize_preds=True`, which you can set to `False` to achieve the behavior you requested? Obviously, it wouldn't work with all the defined writers, but that's not expected either, in those cases you'd just want to use a writer that performs `torch.save()` or `pickle.dump()`.

That way any custom `self.write()` is gonna be simpler, and if your particular `Writer` class always expects to behave as you explained, you just specify `super().__init__(..., atomize_preds=False)`, and do not include that argument in the `__init__` of you `Writer`.

Reply:
> But the multi_pred_id is essentially pred keys right?
Yes, obtained from the `flatten_structure` which flattens dict, lists, etc., as well as any nested version of those, into a flat dict, where keys will be the key or index (or any combo of those in nested) of each pred in the multi-output pred structure.

Reply:
> I find this multi_pred_id very confusing naming. For the id these are some unique identifiers so it makes sense.

`id` identifies pred with regards to the input sample, whereas `multi_pred_id` identifies each individual output in multi-output pred. think of `id` as a row, and each `multi_pred_id` as column in a table.

Do you think we should get rid of this whole thing? And only ever save the pred the way it's outputted?

Reply:
> Do you think we should get rid of this whole thing? And only ever save the pred the way it's outputted?

Yes, 100%

Throughout the framework, we assume that `pred` is a single object that is then handled by the user in model outputs, loss, etc. Doesn't make too much sense imo to change it now and offer this granularity here. 

Of course, the user can save it how they want if we provide it to them directly without assuming this `multi_pred_id`

Reply:
@surajpaib Remove the multi_pred_id, it allowed me to simplify a lot of things as well. 100% agree with it now, too much granularity, let users just override it if they need it, which is probably rare.

Currently, the LighterLogger supports the same kind of feature - should I remove it from there too and keep it simple?